waspc/waspls/src/Wasp/LSP/SignatureHelp.hs

```diff
@@ -0,0 +1,261 @@
+                          _label = Text.pack $ showFragments signatureFragments,
+                          _documentation = Nothing
+                        }
+                    ],
+                _activeSignature = Just 0,
+                _activeParameter = Nothing
+              }
+  where
+    emptyHelp =
+      LSP.SignatureHelp
+        { _signatures = LSP.List [],
+          _activeSignature = Nothing,
+          _activeParameter = Nothing
+        }
+
+-- | 'Signature' describes the expected type at a specific location inside of
+-- a container in a wasp file. Every signature includes the type of the container
+-- surrounding the location. For example, the container type for a cursor
+-- positioned inside of a dictionary would be the type of the dictionary.
+--
+-- When the location is at a more specific place inside of the container--for
+-- example, immediately after @key: @ in a dictionary--then we say the location
+-- is at a parameter, and the parameter is described in relation to the container
+-- with an 'ExprKey'.
+data Signature
+  = -- | Inside a container, but not at any parameter inside of it.
+    InContainer !Type
+  | -- | Inside a container at the parameter specified by the 'ExprKey'. This
+    -- is named \"param\" to be consistent with the naming conventions used by
+    -- the LSP.
+    AtParam !Type !ExprKey
+  deriving (Eq, Show)
+
+signatureType :: Signature -> Type
+signatureType (InContainer typ) = typ
+signatureType (AtParam typ _) = typ
+
+signatureParam :: Signature -> Maybe ExprKey
+signatureParam (InContainer _) = Nothing
+signatureParam (AtParam _ key) = Just key
+
+-- | @'findSignatureAtLocation' sourceCode location@ runs type inference at the
+-- given location to try to find a 'Signature'.
+--
+-- To do this, two types are inferred: the /tip type/ and the /parent type/. The
+-- tip type is the type expected exactly at the cursor location and the parent
+-- type is the type of the parent node of the tip. For example, if the cursor is
+-- at the 2nd entry of a tuple with type @(string, number)@, then the tip type
+-- is @number@ and the parent type is that tuple type.
+--
+-- Then we check if the cursor is 'InContainer' or 'AtParam'. It is 'InContainer'
+-- only when the tip type is a container or if there is no parent type.
+findSignatureAtLocation ::
+  (MonadLog m) =>
+  String ->
+  Traversal ->
+  m (Maybe Signature)
+findSignatureAtLocation src location = runMaybeT $ do
+  exprPath <- hoistMaybe $ findExprPathAtLocation src location
+  lift $ logM $ "[SignatureHelp] at expr path " ++ show exprPath
+  guard $ not $ null exprPath
+  case exprPath of
+    [path] -> InContainer <$> hoistMaybe (findTypeForPath [path])
+    path -> do
+      -- Using init/last here is OK since we know @path@ has at least 2 elements.
+      tipType <- hoistMaybe $ findTypeForPath path
+      parentType <- hoistMaybe $ findTypeForPath $ init path
```

Comment:
Is there a reason why you do this check here before you do `if isContainerType`?
Because there is a branch where you won't need this result. So you could move this line into second branch. Unless you want to shortcircuit here if parentType can't be found.